### PR TITLE
Introduce store-gateway-automated-downscale.libsonnet into operations/mimir

### PR DESCRIPTION
**Pull request by duricanikolic**
#### What this PR does
This PR introduces a possibility to execute automated store-gateway downscaling zone by zone. This feature can be enabled via the `store_gateway_automated_downscale_enabled` flag. The default waiting time between downscaling 

This PR allows store-gateways to be safely downscaled using the `prepare-downscale` and `downscale-leader` features of the rollout-operator. It is nnly available when using multi-zone store-gateways. This feature can be enabled via the `store_gateway_automated_downscale_enabled` flag. When enabled, the number of replicas in zone B will be scaled to match the number of replicas in zone A on a configurable delay. Correspondingly, zone C will follow zone B on the same configurable delay. The default delay can be changed via `$._config.store_gateway_automated_downscale_min_time_between_zones`. 
The number of replicas in zone A is still controlled by `$._config.multi_zone_store_gateway_replicas` until a reasonable metric to scale it automatically is determined and tested.

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: operations/mimir/store-gateway-autoscaling.libsonnet
Comment (pracucci) [inline operations/mimir/store-gateway-autoscaling.libsonnet:14]: [nit] I would rename to match the "enabled" config option name.

```suggestion
    store_gateway_automated_downscale_min_time_between_zones: if $._config.store_gateway_lazy_loading_enabled then '15m' else '60m',
```

File: operations/mimir/store-gateway-autoscaling.libsonnet
Comment (pracucci) [inline operations/mimir/store-gateway-autoscaling.libsonnet:1]: I would rename the file to `store-gateway-automated-downscale.libsonnet` since this is not autoscaling yet.

Comment (pracucci) [inline operations/mimir/store-gateway-autoscaling.libsonnet:1]: Please make a copy of `operations/mimir-tests/test-multi-zone.jsonnet` to something like `operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling.jsonnet`, then enable the downscaling and re-generate the manifests output running `make build-jsonnet-tests`. Then we can check the diff between `operations/mimir-tests/test-multi-zone-generated.yaml` and the new file `operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling-generated.yaml`

**Review by pracucci — APPROVED** [review]
Thanks! LGTM (modulo a couple of nits).

File: operations/mimir/store-gateway-autoscaling.libsonnet
Comment (duricanikolic) [reply operations/mimir/store-gateway-autoscaling.libsonnet:1]: @pracucci 
I was actually wondering why the tests were not re-generated although I previously did `make build-jsonnet-tests`: the reason was that I didn't import `store-gateway-automated-downscaling.libsonnet` in `mimir.libsonnet`. This is now fixed.

I have also added the new test file `operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling.jsonnet`, and in order to make it different from `operations/mimir-tests/test-multi-zone.jsonnet` I changed the default value of `store_gateway_automated_downscale_min_time_between_zones`, since automated downscaling is _enabled by default_.

File: operations/mimir-tests/test-read-write-deployment-mode-s3-generated.yaml
Comment (pracucci) [inline operations/mimir-tests/test-read-write-deployment-mode-s3-generated.yaml:984]: This is interesting. We shouldn't change it here. I think you need to apply the change only when deployment mode is microservices.

File: operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling.jsonnet
Comment (pracucci) [inline operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling.jsonnet:22]: When I left the comment about this test, I completely missed the fact that we're enabling it by default. Then this test doesn't make sense. Please remove it.

Comment (duricanikolic) [reply operations/mimir-tests/test-multi-zone-with-store-gateway-automated-downscaling.jsonnet]: I have renamed the test into `test-multi-zone-without-store-gateway-automated-downscaling.jsonnet` and tested what happen when we disable automated downscaling.

File: operations/mimir-tests/test-consul-multi-zone-generated.yaml
Comment (pracucci) [inline operations/mimir-tests/test-consul-multi-zone-generated.yaml:2373]: I don't understand. Why this change?

File: operations/mimir-tests/test-deployment-mode-migration-generated.yaml
Comment (pracucci) [inline operations/mimir-tests/test-deployment-mode-migration-generated.yaml:2404]: This is interesting. We should not apply the change while the migration is in progress.

File: operations/mimir-tests/test-consul-multi-zone-generated.yaml
Comment (duricanikolic) [reply operations/mimir-tests/test-consul-multi-zone-generated.yaml:2373]: This is done on `StatefulSet`s of `zone-b` and `zone-c`, since their number of replicas will be handled by the rollout operator, while the `StatefulSet` of `zone-a` is handled by kubernetes.

During the migration we also need to flux-ignore the stateful sets before the change is merged.

Comment (pracucci) [reply operations/mimir-tests/test-consul-multi-zone-generated.yaml:2373]: Ok, but the change is unrelated from your PR: Why do we see such diff now?

**Review by pracucci — APPROVED** [review]
LGTM, thanks for your patience!

P.S. I pushed a commit to simplify the test file. I hope you don't mind!

**Pull request merged**
